### PR TITLE
kd-tree

### LINKED ISSUE #15: Implement kdtree structure
For more optimal intersection checking

### PR DESCRIPTION
**Pull request by Tako-San**
Implement kd-tree to speed up triangles intersection. Closes #15

### CODE REVIEW

**Comment by derzhavin3016** [comment]
Replace trivial definitions of copy/move ctors/assign with default ones (`= default`)

**Comment by Tako-San** [comment]
> Replace trivial definitions of copy/move ctors/assign in Container class with default ones (`= default`)

Done

**Review by Tako-San — COMMENTED** [review]
Looks like done. Need to cleanup all this stuff

**Review by Tako-San — COMMENTED** [review]
Get rid of recursion

File: include/kdtree/node.hh [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    ost << reinterpret_cast<std::uintptr_t>(this) << " -> "
+        << reinterpret_cast<std::uintptr_t>(left.get()) << " [label=\"L\"];" << std::endl;
+  }
+  if (right)
+  {
+    right->dumpRecursive(ost);
+    ost << reinterpret_cast<std::uintptr_t>(this) << " -> "
+        << reinterpret_cast<std::uintptr_t>(right.get()) << " [label=\"R\"];" << std::endl;
+  }
+}
```
Comment (Tako-San) [inline include/kdtree/node.hh:56]: Make non-recursive dump

File: include/kdtree/kdtree.hh [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+typename KdTree<T>::ConstIterator::reference KdTree<T>::ConstIterator::operator*() const
+{
+  return Container<T>{tree_, node_};
+}
+
+// template <std::floating_point T>
+// typename KdTree<T>::ConstIterator::pointer KdTree<T>::ConstIterator::operator->() const
+// {
+//   return std::make_unique<Container<T>>(tree_, node_);
+// }
```
Comment (Tako-San) [inline include/kdtree/kdtree.hh:431]: Handle operator-> or remove it

Comment (Tako-San) [reply include/kdtree/kdtree.hh]: Done

File: tools/lvl1/main.cc [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+  if (KdTree<T>::isOnPosSide(node.sepAxis(), node.separator(), *trIt))
+    intersectHelper(trIt, intInd, node.right());
+  else if (KdTree<T>::isOnNegSide(node.sepAxis(), node.separator(), *trIt))
+    intersectHelper(trIt, intInd, node.left());
+  else if (node.sepAxis() != Axis::NONE)
+  {
+    intersectHelper(trIt, intInd, node.right());
+    intersectHelper(trIt, intInd, node.left());
+  }
```
[the remark this answers is not in the document]
Comment (Tako-San) [reply tools/lvl1/main.cc]: Done

File: include/kdtree/kdtree.hh [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+    triangles_.push_back(tr);
+    nonExpandingInsert(root_.get(), tr, index);
+  }
+}
+
+template <std::floating_point T>
+void KdTree<T>::clear()
+{
+  // Temporary recursive solution
+  root_.reset();
```
[the remark this answers is not in the document]
Comment (Tako-San) [reply include/kdtree/kdtree.hh]: Done

**Review by derzhavin3016 — APPROVED** [review]
Nice

File: include/kdtree/kdtree.hh [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+typename KdTree<T>::ConstIterator::reference KdTree<T>::ConstIterator::operator*() const
+{
+  return Container<T>{tree_, node_};
+}
+
+template <std::floating_point T>
+typename KdTree<T>::ConstIterator::pointer KdTree<T>::ConstIterator::operator->() const
+{
+  return {tree_, node_};
+}
```
[the remark this answers is not in the document]
Comment (Tako-San) [reply include/kdtree/kdtree.hh]: Outdated

**Pull request merged**